### PR TITLE
Validate host routes

### LINKED ISSUE #439: Docker0 bridge can conflict with service/pod cidrs
This is pretty hard to spot (at least with calico)... everything else works but http liveness/readiness probes will fail.

### PR DESCRIPTION
**Pull request by SpComb**
Fixes #439

Fail if any of the non-network-provider-managed routes on the host overlaps with any of the kube network routes (`pod_network_cidr`, `service_cidr`).

### CODE REVIEW

**Comment by SpComb** [comment]
Specs pending design ACK.

**Comment by SpComb** [comment]
Example with calico:

```
    $ ip route
    default via 167.99.32.1 dev eth0 onlink 
    blackhole 10.32.0.0/24  proto bird 
    10.32.0.35 dev cali73282f4333c  scope link 
    10.32.0.36 dev cali6f15a1d419d  scope link 
    10.32.0.37 dev cali5f05af94801  scope link 
    10.32.0.38 dev cali5f1ddd73716  scope link 
    10.32.1.0/24 via 206.189.0.173 dev tunl0  proto bird onlink 
    10.32.2.0/24 via 188.166.118.151 dev tunl0  proto bird onlink 
    10.32.3.0/24 via 167.99.139.236 dev tunl0  proto bird onlink 
    167.99.32.0/20 dev eth0  proto kernel  scope link  src 167.99.39.233 
    172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown 
    ! 0
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="default via 167.99.32.1 dev eth0 onlink" type=nil prefix="0.0.0.0/0" via="167.99.32.1" dev="eth0" proto=nil options="onlink">
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="blackhole 10.32.0.0/24  proto bird" type="blackhole" prefix="10.32.0.0/24" via=nil dev=nil proto="bird" options=nil>
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="10.32.0.35 dev cali73282f4333c  scope link" type=nil prefix="10.32.0.35" via=nil dev="cali73282f4333c" proto=nil options="scope link">
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="10.32.0.36 dev cali6f15a1d419d  scope link" type=nil prefix="10.32.0.36" via=nil dev="cali6f15a1d419d" proto=nil options="scope link">
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="10.32.0.37 dev cali5f05af94801  scope link" type=nil prefix="10.32.0.37" via=nil dev="cali5f05af94801" proto=nil options="scope link">
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="10.32.0.38 dev cali5f1ddd73716  scope link" type=nil prefix="10.32.0.38" via=nil dev="cali5f1ddd73716" proto=nil options="scope link">
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="10.32.1.0/24 via 206.189.0.173 dev tunl0  proto bird onlink" type=nil prefix="10.32.1.0/24" via="206.189.0.173" dev="tunl0" proto="bird" options="onlink">
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="10.32.2.0/24 via 188.166.118.151 dev tunl0  proto bird onlink" type=nil prefix="10.32.2.0/24" via="188.166.118.151" dev="tunl0" proto="bird" options="onlink">
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="10.32.3.0/24 via 167.99.139.236 dev tunl0  proto bird onlink" type=nil prefix="10.32.3.0/24" via="167.99.139.236" dev="tunl0" proto="bird" options="onlink">
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="167.99.32.0/20 dev eth0  proto kernel  scope link  src 167.99.39.233" type=nil prefix="167.99.32.0/20" via=nil dev="eth0" proto="kernel" options="scope link  src 167.99.39.233">
    [167.99.39.233] ip route: #<Pharos::Configuration::Host::Route raw="172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown" type=nil prefix="172.17.0.0/16" via=nil dev="docker0" proto="kernel" options="scope link  src 172.17.0.1 linkdown">
    [167.99.39.233] Validating current role matches ...
    [167.99.39.233] master role matches
    [167.99.39.233] Validating distro and version ...
    [167.99.39.233] Validating host configuration ...
    [167.99.39.233] Validating host routes ...
 [Validate hosts @ 167.99.39.233] RuntimeError: Overlapping host routes for .network.pod_network_cidr=172.16.0.0/12: 172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown
Overlapping host routes for .network.pod_network_cidr=172.16.0.0/12: 172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown
/app/lib/pharos/phases/validate_host.rb:206:in `validate_routes'
/app/lib/pharos/phases/validate_host.rb:22:in `call'
/app/lib/pharos/phase_manager.rb:70:in `block in apply'
/app/lib/pharos/phase_manager.rb:26:in `block (2 levels) in run_parallel'
```

File: lib/pharos/phases/validate_host.rb [hunk 1]
```diff
@@ -163,6 +166,50 @@ def read_resolvconf
           systemd_resolved_stub: check_resolvconf_systemd_resolved_stub
         )
       end
+
+      ROUTE_REGEXP = %r(^((?<type>\S+)\s+)?(?<prefix>default|[0-9./]+)(\s+via (?<via>\S+))?(\s+dev (?<dev>\S+))?(\s+proto (?<proto>\S+))?(\s+(?<options>.+))?$)
+
+      # @return [Array<Pharos::Configuration::Host::Route>]
+      def read_routes
```
Comment (jakolehm) [inline lib/pharos/phases/validate_host.rb]: `ValidateHost` starts to get bloated.. maybe wrap these helpers to a separate class?

Comment (SpComb) [reply lib/pharos/phases/validate_host.rb]: I can move the parsing to `Pharos::Configuartion::Host::Route#parse`... would that be enough?

Comment (SpComb) [reply lib/pharos/phases/validate_host.rb]: Could also be a separate Phase.

Comment (jakolehm) [reply lib/pharos/phases/validate_host.rb]: I would prefer to have just single validation phase and split the logic to separate classes. `Pharos::Configuartion::Host::Route#parse` fits that bill.

File: lib/pharos/configuration/host.rb [hunk 2]
```diff
@@ -13,6 +13,46 @@ class ResolvConf < Dry::Struct
         attribute :systemd_resolved_stub, Pharos::Types::Strict::Bool
       end
 
+      class Route < Dry::Struct
```
Comment (jakolehm) [inline lib/pharos/configuration/host.rb]: Inner classes are usually considered as private to the outer class. In this case `Host::Route` is called from outside and it does not feel right.

Comment (jakolehm) [reply lib/pharos/configuration/host.rb]: Right, I didn't notice those other embedded classes from the diff. Maybe `Pharos::Configuration::HostRoute` makes more sense but we can also rethink this later (visual/coding-style issue).

**Pull request merged**